### ISSUE TITLE
Why require a fieldseparator after a null value?

### ISSUE THREAD

**Issue by RobThree** [Stenway/RSV-Challenge#2]
Why would you require a field separator (`0xFF`) after a `null` value (`0xFE`)? Why not have `0xFE` signal _both_ the null value _and_ "end of field"? That would make reading RSV much easier because when reading a `0xFE` you can immediately emit the `null` value and don't need any more shenanigans to wait for reading the `0xFF`, looking back in the buffer if there's one and only one byte in it and it is `null` etc.

**Comment by txgruppi** [comment]
I would get right of `0xFE` and identify a null value by a zero length byte sequence read from the previous to the next `0xFF`.

This can simplify the spec by having only 2 special bytes:

- `0xFF` for `End Of Value`
- `0xFE` for `End of Row`

**Comment by RobThree** [comment]
No, that won't work because you won't be able to store an empty string (which is not the same as a null value).

**Comment by Osndok** [comment]
I agree with Rob's suggestion and would like to add one additional reason. I would like to interpret 0xFF as "end of value that is present", and 0xFE as "end of value that is not-present/disabled". The algorithm the becomes to "read bytes up to an end mark" in either case (it's just the current case is that all nulls are empty strings, but loops in my intended use case (or maybe it is a debugging use-case?) where values can be toggled on/off by changing one byte (changing 0xFF to 0xFE) while keeping the file otherwise unmodified.

I understand it is probably outside of the original scope of RSV, but just thought it would be better to speak up now and potentially avoid a confusion-inducing RSV fork for our oddball case.

**Comment by dcoai** [comment]
I think having multiple separators complicates decoding and is a minor optimization for most use cases.  The way it is, in some languages you can use split(string, 0xFD) or split(string, 0xFF) to split the rows and fields and use iterators to operate on the results... this is nice.

**Comment by RobThree** [comment]
A `split()` will typically require you to pass the entire data which makes it horribly inefficiënt and, essentially, limits you to whatever memory you have before having to hit the swapfile etc. A split is cool for simple files, but as they get larger you'll want a reader that is streaming. That way you can read bytes and, as you encounter a separator or `null` value, emit each field / row. Also, without a `null` indicator you won't be able to differentiate between `null` and an empty string, which is, again, a huge limiting factor.

**Comment by dcoai** [comment]
Very true regarding streaming, though split could be set up to stream as well.  I haven't examined very many examples in this repo, but the ones I have don't look to be set up for streaming.

What I was getting at and didn't explain well is that adding decisions in the code adds complexity to the decoding processes.  My understanding of the point of RSV was simplicity.  I realize it is not great complexity that it adds, but those sorts of decisions can make an elegant solution much less so quickly.

**Comment by RobThree** [comment]
@dcoai I'm still not sure I understand. You're advocating keeping the field separator after a null value instead of what I proposed in this issue (getting rid of the field separator after a null value)? Because, yes, that would make sense when using `split()` etc. and I can see that point. But you don't address what happens when files get bigger and where (common) `split()` implementations will run out of memory as opposed to a streaming reader which will have, essentially, no limitations.

I think also the (common) `split()` will suffer a lot more from [Shlemiel the painter’s problem](https://www.joelonsoftware.com/2001/12/11/back-to-basics/) and be much slower because of many (re)allocations where a streaming algorithm can just start reading and emit fields and rows as it goes. I'm not saying a streaming `split()` can't be done, but I _am_ saying that in most languages it isn't as far as I'm aware.

**Comment by RobThree** [comment]
@Stenway 
Maybe this issue should be transferred to https://github.com/Stenway/RSV-Specification now that the spec has a separate repository?

**Comment by Stenway** [comment]
Hi there,
thanks for your discussion about this. The null-byte terminator was also a hot topic in the comments section of the video, so I've added a frequently asked questions section (FAQ) to the spec repository: 
https://github.com/Stenway/RSV-Specification#why-is-the-null-value-byte-terminated-with-a-value-terminator-byte
also concerning the differentiation between empty strings and null values:
https://github.com/Stenway/RSV-Specification#why-does-rsv-differentiate-between-empty-strings-and-null-values

Split is definitely not ideal. And sometimes there isn't a builtin byte-array based split method in a programming language. But when it is available, it usually leads to less lines of code for the decoder. I've implemented versions in Python, Go and C# for testing purposes. The Python implementation is actually quite nice I think: https://github.com/Stenway/RSV-Challenge/blob/main/Python/rsv.py#L70

So sacrificing this might harm the adoption.

Nevertheless I've added an Experiments folder to the repository, and added an implementation without the terminator byte for the null byte, for comparison:
https://github.com/Stenway/RSV-Challenge/tree/main/Experiments/NullByteWithoutTerminator

Cheers
Stefan
